onnxscript/rewriter/llama_rule_sets_test.py

```diff
@@ -127,6 +130,217 @@ def test_llama_p0_rule_set_transpose_transpose(self):
+            ir_model = ir.serde.deserialize_model(model_proto)
+            rule_set = llama_rule_sets.llama_p0_rule_set()
+            rule_set.apply_to_model(ir_model)
+            rewritten_model = ir.serde.serialize_model(ir_model)
+
+            self.assertEqual(
+                ["Constant", "Unsqueeze"], [n.op_type for n in rewritten_model.graph.node]
+            )
+            self._check_model(model_proto, rewritten_model)
+
+    @classmethod
+    def _reshape_reshape_models(cls):
+        models = [
+            onnx.helper.make_model(
+                onnx.helper.make_graph(
+                    [
+                        onnx.helper.make_node("Reshape", ["X", "shape_"], ["Xu"]),
+                        onnx.helper.make_node("Reshape", ["Xu", "shape"], ["Y"]),
+                    ],
+                    "name",
+                    [onnx.helper.make_tensor_value_info("X", FLOAT, [3, 4, 5])],
+                    [onnx.helper.make_tensor_value_info("Y", FLOAT, [5, 4, 3])],
+                    [
+                        onnx.numpy_helper.from_array(
+                            np.array([4, 5, 3], dtype=np.int64), name="shape_"
+                        ),
+                        onnx.numpy_helper.from_array(
+                            np.array([5, 4, 3], dtype=np.int64), name="shape"
+                        ),
+                    ],
+                ),
+                opset_imports=[onnx.helper.make_opsetid("", 18)],
+            ),
+            onnx.helper.make_model(
+                onnx.helper.make_graph(
+                    [
+                        onnx.helper.make_node("Reshape", ["X", "shape_"], ["Xu"]),
+                        onnx.helper.make_node("Reshape", ["Xu", "shape"], ["Y"]),
+                    ],
+                    "name",
+                    [onnx.helper.make_tensor_value_info("X", FLOAT, [3, 4, 5])],
+                    [onnx.helper.make_tensor_value_info("Y", FLOAT, [5, 4, 3])],
+                    [
+                        onnx.numpy_helper.from_array(
+                            np.array([-1], dtype=np.int64), name="shape_"
+                        ),
+                        onnx.numpy_helper.from_array(
+                            np.array([5, 4, 3], dtype=np.int64), name="shape"
+                        ),
+                    ],
+                ),
+                opset_imports=[onnx.helper.make_opsetid("", 18)],
+            ),
+        ]
+        return models
+
+    def test_llama_p0_rule_set_reshape_reshape(self):
+        for model_proto in self._reshape_reshape_models():
+            ir_model = ir.serde.deserialize_model(model_proto)
+            rule_set = llama_rule_sets.llama_p0_rule_set()
+            rule_set.apply_to_model(ir_model)
+            rewritten_model = ir.serde.serialize_model(ir_model)
+
+            self.assertEqual(
+                ["Reshape"], [n.op_type for n in rewritten_model.graph.node]
+            )
```

Comment:
I would leverage the IR for comparisons and avoid checks on the proto unless necessary.

Reply:
The condition applies on the proto because this is what we use at the end. Not sure it would be better with the ir.

Reply:
We should be confident that checking the IR _is_ checking the proto; this way the proto is only used as a serialization format and nothing more.

Reply:
`["Reshape"], [n.op_type for n in ir_model.graph]`